### ISSUE TITLE
microseconds being retrieved incorrectly with NonStop SQL

### ISSUE THREAD

**Issue by bhaskar219** [mkleehammer/pyodbc#908]
Hi all the datetime data for the microseconds retrieving wrongly, So getting wrong results.
for ex: first record microsecond value is 521243 is  appeared as 521 , so finally appeared as 000521
So finally 521243   appeared as 000521 , So wrong results happening. 
I verified in SQLMX database .  not sure where is the problem . when i verified through odbc program received correct results. but through pyodbc getting wrong results

This is correct data 
>>> for i in cur1.execute('''select cast(FILE_DATE as char(26)),FILE_ID from t1 where FILE_ID='0791POC001' '''):
...     i
... 
('2011-03-15 18:35:59.**521243**', '0791POC001')
('2011-03-23 06:20:37.615464', '0791POC001')
('2011-03-20 08:10:54.678461', '0791POC001')
('2011-03-15 03:41:50.675359', '0791POC001')
('2011-03-25 01:01:28.496676', '0791POC001')
('2011-03-24 14:41:38.217177', '0791POC001')
('2011-03-21 05:07:41.149088', '0791POC001')
('2011-03-14 23:44:27.447464', '0791POC001')
('2011-03-21 18:23:49.313502', '0791POC001')
('2011-03-17 05:39:47.575087', '0791POC001')
('2011-03-23 06:26:18.383204', '0791POC001')

# The microseconds values truncated wrongly
>>> for i in cur1.execute('''select FILE_DATE ,FILE_ID from t1 where FILE_ID='0791POC001' '''):    
...     i
... 
(datetime.datetime(2011, 3, 15, 18, 35, 59, **521**), '0791POC001')
(datetime.datetime(2011, 3, 23, 6, 20, 37, 615), '0791POC001')
(datetime.datetime(2011, 3, 20, 8, 10, 54, 678), '0791POC001')
(datetime.datetime(2011, 3, 15, 3, 41, 50, 675), '0791POC001')
(datetime.datetime(2011, 3, 25, 1, 1, 28, 496), '0791POC001')
(datetime.datetime(2011, 3, 24, 14, 41, 38, 217), '0791POC001')
(datetime.datetime(2011, 3, 21, 5, 7, 41, 149), '0791POC001')
(datetime.datetime(2011, 3, 14, 23, 44, 27, 447), '0791POC001')
(datetime.datetime(2011, 3, 21, 18, 23, 49, 313), '0791POC001')
(datetime.datetime(2011, 3, 17, 5, 39, 47, 575), '0791POC001')
(datetime.datetime(2011, 3, 23, 6, 26, 18, 383), '0791POC001')

# the values changed from 521242 to 000521
>>> for i in cur1.execute('''select FILE_DATE ,FILE_ID from t1 where FILE_ID='0791POC001' '''):
...     i[0].strftime("%d-%b-%Y (%H:%M:%S.%f)"),i[1]
... 
('15-Mar-2011 (18:35:59.**000521**)', '0791POC001')
('23-Mar-2011 (06:20:37.000615)', '0791POC001')
('20-Mar-2011 (08:10:54.000678)', '0791POC001')
('15-Mar-2011 (03:41:50.000675)', '0791POC001')
('25-Mar-2011 (01:01:28.000496)', '0791POC001')
('24-Mar-2011 (14:41:38.000217)', '0791POC001')
('21-Mar-2011 (05:07:41.000149)', '0791POC001')
('14-Mar-2011 (23:44:27.000447)', '0791POC001')
('21-Mar-2011 (18:23:49.000313)', '0791POC001')
('17-Mar-2011 (05:39:47.000575)', '0791POC001')
('23-Mar-2011 (06:26:18.000383)', '0791POC001')
>>>

**Comment by gordthompson** [comment]
Does the reference to "SQLMX" mean that you are using a [NonStop SQL](https://en.wikipedia.org/wiki/NonStop_SQL) database?

**Comment by bhaskar219** [comment]
> Does the reference to "SQLMX" mean that you are using a [NonStop SQL](https://en.wikipedia.org/wiki/NonStop_SQL) database?

yes

**Comment by gordthompson** [comment]
This is not a problem with pyodbc *per se*. You are probably encountering a deficiency in the ODBC driver. The following code works for me with SQL Server:

```python
import datetime
import platform
import sys

import pyodbc

print(f"platform: {platform.system()}")
sys_version = sys.version.replace("\n", "")
print(f"Python version: {sys_version}")
print(f"pyodbc version: {pyodbc.version}")

cnxn = pyodbc.connect(
    "DRIVER=ODBC Driver 17 for SQL Server;"
    "SERVER=192.168.0.199;"
    "DATABASE=test;"
    "UID=scott;PWD=tiger^5HHH;"
)
print(
    f"ODBC driver: {cnxn.getinfo(pyodbc.SQL_DRIVER_NAME)}"
    f" ({cnxn.getinfo(pyodbc.SQL_DRIVER_VER)})"
)
cnxn.autocommit = True
crsr = cnxn.cursor()

table_name = "gh_908"
crsr.execute(f"DROP TABLE IF EXISTS {table_name}")
crsr.execute(f"CREATE TABLE {table_name} (id int primary key, dt datetime)")

dt_str = "2011-03-24 14:41:31.217177"
dt_value = datetime.datetime.strptime(dt_str, "%Y-%m-%d %H:%M:%S.%f")
print(f"value stored: {repr(dt_value)}")
crsr.execute(f"INSERT INTO {table_name} (id, dt) VALUES (1, ?)", dt_value)

actual_value = crsr.execute(f"SELECT dt FROM {table_name} WHERE id = 1").fetchval()
print(f"value retrieved: {repr(actual_value)}")
expected_value = datetime.datetime(2011, 3, 24, 14, 41, 31, 217000)
print(f"expected value: {repr(expected_value)}")
assert actual_value == expected_value
print("Test passed.")


"""console output:
platform: Linux
Python version: 3.8.5 (default, May 27 2021, 13:30:53) [GCC 9.3.0]
pyodbc version: 4.0.31b56
ODBC driver: libmsodbcsql-17.7.so.2.1 (17.07.0002)
value stored: datetime.datetime(2011, 3, 24, 14, 41, 31, 217177)
value retrieved: datetime.datetime(2011, 3, 24, 14, 41, 31, 217000)
expected value: datetime.datetime(2011, 3, 24, 14, 41, 31, 217000)
Test passed.
"""
```

You should follow up with the people who maintain the ODBC driver that you are using.

**Comment by bhaskar219** [comment]
ok  I will check with them and get back  with updates

**Comment by bhaskar219** [comment]
value stored: datetime.datetime(2011, 3, 24, 14, 41, 31, 217177)
value retrieved: datetime.datetime(2011, 3, 24, 14, 41, 31, 217000)
expected value: datetime.datetime(2011, 3, 24, 14, 41, 31, 217000)

Why last three digits are not captured  and it appeared as Zero ? 
217177   --- > 217000 
This is also wrong right ?

**Comment by bhaskar219** [comment]
> This is not a problem with pyodbc _per se_. You are probably encountering a deficiency in the ODBC driver. The following code works for me with SQL Server:
> 
> ```python
> import datetime
> import platform
> import sys
> 
> import pyodbc
> 
> print(f"platform: {platform.system()}")
> sys_version = sys.version.replace("\n", "")
> print(f"Python version: {sys_version}")
> print(f"pyodbc version: {pyodbc.version}")
> 
> cnxn = pyodbc.connect(
>     "DRIVER=ODBC Driver 17 for SQL Server;"
>     "SERVER=192.168.0.199;"
>     "DATABASE=test;"
>     "UID=scott;PWD=tiger^5HHH;"
> )
> print(
>     f"ODBC driver: {cnxn.getinfo(pyodbc.SQL_DRIVER_NAME)}"
>     f" ({cnxn.getinfo(pyodbc.SQL_DRIVER_VER)})"
> )
> cnxn.autocommit = True
> crsr = cnxn.cursor()
> 
> table_name = "gh_908"
> crsr.execute(f"DROP TABLE IF EXISTS {table_name}")
> crsr.execute(f"CREATE TABLE {table_name} (id int primary key, dt datetime)")
> 
> dt_str = "2011-03-24 14:41:31.217177"
> dt_value = datetime.datetime.strptime(dt_str, "%Y-%m-%d %H:%M:%S.%f")
> print(f"value stored: {repr(dt_value)}")
> crsr.execute(f"INSERT INTO {table_name} (id, dt) VALUES (1, ?)", dt_value)
> 
> actual_value = crsr.execute(f"SELECT dt FROM {table_name} WHERE id = 1").fetchval()
> print(f"value retrieved: {repr(actual_value)}")
> expected_value = datetime.datetime(2011, 3, 24, 14, 41, 31, 217000)
> print(f"expected value: {repr(expected_value)}")
> assert actual_value == expected_value
> print("Test passed.")
> 
> 
> """console output:
> platform: Linux
> Python version: 3.8.5 (default, May 27 2021, 13:30:53) [GCC 9.3.0]
> pyodbc version: 4.0.31b56
> ODBC driver: libmsodbcsql-17.7.so.2.1 (17.07.0002)
> value stored: datetime.datetime(2011, 3, 24, 14, 41, 31, 217177)
> value retrieved: datetime.datetime(2011, 3, 24, 14, 41, 31, 217000)
> expected value: datetime.datetime(2011, 3, 24, 14, 41, 31, 217000)
> Test passed.
> """
> ```
> 
> You should follow up with the people who maintain the ODBC driver that you are using.

value stored: datetime.datetime(2011, 3, 24, 14, 41, 31, 217177)
value retrieved: datetime.datetime(2011, 3, 24, 14, 41, 31, 217000)
expected value: datetime.datetime(2011, 3, 24, 14, 41, 31, 217000)

Why last three digits are not captured and it appeared as Zero ?
217177 --- > 217000
This is also wrong right ?

Why capturing only 3 digits to get microseconds  ?

**Comment by gordthompson** [comment]
> Why last three digits are not captured and it appeared as Zero ?
> 217177 --- > 217000
> This is also wrong right ?
> 
> Why capturing only 3 digits to get microseconds ?

That is expected behaviour for a Microsoft SQL Server [datetime](https://docs.microsoft.com/en-us/sql/t-sql/data-types/datetime-transact-sql?view=sql-server-ver15) column. It only stores three digits of fractional seconds. In fact, it does not even support full millisecond precision; it rounds the time portion to the nearest 1/300 of a second.

**Comment by rajaraokg1** [comment]
Hi,

I debuged with simple ODBC application through NonStop ODBC driver and found that SQLGetData with ctype as SQL_C_TYPE_TIMESTAMP and able to return proper timestamp structure data.

Snipset of the application code:
```
st = SQLGetData(hstmt, icol, SQL_C_TYPE_TIMESTAMP, &timestampData, 0, &cbOutputLen[icol-1]);
sprintf (strColData, "%d-%d-%d %d:%d:%d.%d \t", timestampData.year, timestampData.month, timestampData.day, timestampData.hour, timestampData.minute, timestampData.second, timestampData.fraction);
```

Output:
(EXPR)
 2021-6-7 20:6:38.386796000

I want to debug further pyodbc code to figure out what problem we are facing with NonStop SQL. Please help in creating the debug environment for pyodbc to resolve this issue. 

Please share me the debug environment steps for pyodbc.

Thanks & Regards,
Raja.

**Comment by gordthompson** [comment]
Please [enable ODBC tracing](https://github.com/mkleehammer/pyodbc/wiki/Troubleshooting-%E2%80%93-Generating-an-ODBC-trace-log), run a simple Python script that retrieves a date/time value, then attach the ODBC trace log here. (Those logs can be quite long, so please do not just paste it into the text of a comment.)

**Comment by rajaraokg1** [comment]
Yes we did that. Please find the attached Odbc_Trace.log for your reference.
We used the following SQL to see the current timestamp,
`select current from (values(1))t;`

[Odbc_Trace.log](https://github.com/mkleehammer/pyodbc/files/6609644/Odbc_Trace.log)

**Comment by gordthompson** [comment]
What do you see if your Python script does this:

```python
sql = """\
SELECT 
    current AS now_datetime,
    CAST(current AS VARCHAR(50)) AS now_string
FROM (VALUES(1))t;
"""
print(crsr.execute(sql).fetchone())
```

**Comment by bhaskar219** [comment]
> ```python
> print(crsr.execute(sql).fetchone())
> ```

Hi , 
We executed above query and received the following output, 

```
sql='''select current as now_datetime, CAST(current as varchar(50)) as now_string from (values(1))t;'''
print(cur1.execute(sql).fetchone())
```

**output is :**
(datetime.datetime(2021, 6, 8, 12, 4, 27, 617), '2021-06-08 12:04:27.617497')

**Comment by gordthompson** [comment]
Either the database server is sending the fractional seconds incorrectly or the ODBC driver is misinterpreting the value. According to the ODBC specification's [C Data Types](https://docs.microsoft.com/en-us/sql/odbc/reference/appendixes/c-data-types?view=sql-server-ver15), `SQL_C_TYPE_TIMESTAMP` is defined as

```
struct tagTIMESTAMP_STRUCT {  
   SQLSMALLINT year;  
   SQLUSMALLINT month;  
   SQLUSMALLINT day;  
   SQLUSMALLINT hour;  
   SQLUSMALLINT minute;  
   SQLUSMALLINT second;  
   SQLUINTEGER fraction;[b]   
} TIMESTAMP_STRUCT;[a]
```

where note [b] is

> [b] The value of the fraction field is the number of billionths of a second and ranges from 0 through 999,999,999 (1 less than 1 billion). For example, the value of the fraction field for a half-second is 500,000,000, for a thousandth of a second (one millisecond) is 1,000,000, for a millionth of a second (one microsecond) is 1,000, and for a billionth of a second (one nanosecond) is 1.

If the database wants to truncate (or round) the fractional seconds to three decimal places it still has to send the value in nanoseconds, e.g. `617000000`, which pyodbc would then convert to milliseconds: `datetime.datetime(2021, 6, 8, 12, 4, 27, 617000)`.

TL;DR - The NonStop SQL people need to fix this.

**Comment by rajaraokg1** [comment]
Thanks for your support and quick response on clarifying the detail of timestamp more granularly. 

Now I understood the problem. By default our NonStop database adds the fraction part in Micro seconds only. There is an option to change micro seconds to nano seconds by specifying SQLFractionValue attribute while creating the ODBC data source or as part of the connection string attributes. I modified the data source attributes to nano seconds it started giving correct results.

```
con_sqlmx=pyodbc.connect(DSN='TPINT',UID=username,PWD=password,ansi=True,autocommit=True,SQLFractionValue=9)
cur1=con_sqlmx.cursor()
sql='''select current as now_datetime, CAST(current as varchar(50)) as now_string from (values(1))t;'''
print(cur1.execute(sql).fetchone())
con_sqlmx.close()
```

Output:
(datetime.datetime(2021, 6, 9, 21, 12, 31, 807494), '2021-06-09 21:12:31.807494')
